### ISSUE TITLE
Invoke-Pester -PassThru returns pipeline output from test scripts

### ISSUE THREAD

**Issue by mikeclayton** [pester/Pester#1060]
## 1. Provide a general summary of the issue in the Title above

When calling Invoke-Pester with the -PassThru switch, the return value contains all of the pipeline output from the test scripts, as well as the actual test results.

## 2. Describe Your Environment

Operating System, Pester version and PowerShell version:

```
Pester version     : 3.4.0 C:\Program Files\WindowsPowerShell\Modules\Pester\3.4.0\Pester.psd1
PowerShell version : 5.1.16299.251
OS version         : Microsoft Windows NT 10.0.16299.0```
```

Note: I'm actually importing Pester 4.3.1 from a separate working directory, not the version 3.4.0 installed in the PowerShell modules folder.

## 3. Expected Behavior

I'm not sure if this is actually a bug or not, but my expected behaviour would be for Pester to swallow any pipeline output from the test scripts and just return its test results.

## 4.Current Behavior

Create a line-of-business script which sends values to the pipeline when executed. I try to avoid doing this myself, but I've inherited a script which already does it :-(

For example:

**Invoke-MyFunction.ps1**
```powershell
Import-Module -Name "C:\Windows\System32\WindowsPowerShell\v1.0\Modules\BitsTransfer\BitsTransfer.psd1" -PassThru;

Write-Output "my output";

function Invoke-MyFunction
{
    return $true;
}
```

Create a test script:

**Invoke-MyFunction.Tests.ps1**
```powershell
. .\Invoke-MyFunction.ps1;

Describe "MyFunction Tests" {
    It "Should not throw" {
        Invoke-MyFunction
    }
}
```

And a test runner script:

**Invoke-MyPesterTests.ps1**
```powershell
Import-Module -Name "C:\packages\Pester.4.3.1\tools\Pester.psd1" -ErrorAction "Stop";

$results = Invoke-Pester -Script ".\Invoke-MyFunction.Tests.ps1" -PassThru;

write-host "results :";
write-host ($results.GetType().FullName);
write-host ($results | fl * | out-string);
```

Then, run Invoke-MyPesterTests.ps1
```
c:\src\scratch\pester>powershell c:\src\scratch\pester\Invoke-MyPesterTests.ps1
Executing all tests in '.\Invoke-MyFunction.Tests.ps1'

Executing script .\Invoke-MyFunction.Tests.ps1

  Describing MyFunction Tests
    [+] Should call write-output 510ms
Tests completed in 510ms
Tests Passed: 1, Failed: 0, Skipped: 0, Pending: 0, Inconclusive: 0
results :
System.Object[]


LogPipelineExecutionDetails : False
Name                        : BitsTransfer
Path                        : C:\Windows\System32\WindowsPowerShell\v1.0\Modules\BitsTransfer\BitsTransfer.psd1
ImplementingAssembly        :
Definition                  :
Description                 :
Guid                        : 8fa5064b-8479-4c5c-86ea-0d311fe48875
HelpInfoUri                 : https://go.microsoft.com/fwlink/?linkid=390756
ModuleBase                  : C:\Windows\System32\WindowsPowerShell\v1.0\Modules\BitsTransfer
PrivateData                 :
Tags                        : {}
ProjectUri                  :
IconUri                     :
LicenseUri                  :
ReleaseNotes                :
RepositorySourceLocation    :
Version                     : 2.0.0.0
ModuleType                  : Manifest
Author                      : Microsoft Corporation
AccessMode                  : ReadWrite
ClrVersion                  : 4.0
CompanyName                 : Microsoft Corporation
Copyright                   : © Microsoft Corporation. All rights reserved.
DotNetFrameworkVersion      :
ExportedFunctions           : {}
Prefix                      :
ExportedCmdlets             : {[Add-BitsFile, Add-BitsFile], [Complete-BitsTransfer, Complete-BitsTransfer], [Get-BitsTransfer, Get-BitsTransfer], [Remove-BitsTransfer, Remove-BitsTransfer]...}
ExportedCommands            : {[Add-BitsFile, Add-BitsFile], [Complete-BitsTransfer, Complete-BitsTransfer], [Get-BitsTransfer, Get-BitsTransfer], [Remove-BitsTransfer, Remove-BitsTransfer]...}
FileList                    : {}
CompatiblePSEditions        : {}
ModuleList                  : {}
NestedModules               : {Microsoft.BackgroundIntelligentTransfer.Management}
PowerShellHostName          :
PowerShellHostVersion       :
PowerShellVersion           : 3.0
ProcessorArchitecture       : None
Scripts                     : {}
RequiredAssemblies          : {C:\Windows\System32\WindowsPowerShell\v1.0\Modules\BitsTransfer\Microsoft.BackgroundIntelligentTransfer.Management.Interop.dll}
RequiredModules             : {}
RootModule                  :
ExportedVariables           : {}
ExportedAliases             : {}
ExportedWorkflows           : {}
ExportedDscResources        : {}
SessionState                : System.Management.Automation.SessionState
OnRemove                    :
ExportedFormatFiles         : {C:\Windows\System32\WindowsPowerShell\v1.0\Modules\BitsTransfer\BitsTransfer.Format.ps1xml}
ExportedTypeFiles           : {}

my output
TagFilter         : {}
ExcludeTagFilter  : {}
TestNameFilter    :
TotalCount        : 1
PassedCount       : 1
FailedCount       : 0
SkippedCount      : 0
PendingCount      : 0
InconclusiveCount : 0
Time              : 00:00:00.5108241
TestResult        : {@{ErrorRecord=; ParameterizedSuiteName=; Describe=MyFunction Tests; Parameters=System.Collections.Specialized.OrderedDictionary; Passed=True; Show=All; FailureMessage=; Time=00:00:00.5108241; Name=Should call write-output; Result=Passed; Context=; StackTrace=}}

c:\src\scratch\pester>
```

Note that the return value is of type "System.Object[]", and contains the output from Import-Module and Write-Output in the Invoke-MyFunction.ps1 file.

The same effect is also seen if you move these two lines into the Invoke-MyFunction.Tests.ps1 instead.

## 5. Possible Solution

Workarounds are:

* Obviously, don't generate uncaptured output in the test scripts! Use ```[void] ...``` or ```$null = ...``` to capture output

* Select the *last* item in the return value from Invoke-Pester and use that as the expected test results

```powershell
$allOutput = Invoke-Pester -Script ".\Invoke-MyFunction.Tests.ps1" -PassThru;
$testResults = $allOutput | select-object -Last 1;
```

* Modify Invoke-Pester so it only returns the actual test results

I'm not sure if this is actually a bug though, so maybe the second option is the best, given that I can't change the actual test scripts (they're owned by a different team in our company, so I can't change what gets output by the scripts or the tests).

**Comment by mikeclayton** [comment]
I'm happy to do some investigation into this if you can confirm whether it's a bug or not. I didn't want to dig too far into the root cause yet though in case its "by design" behaviour.

**Comment by dlwyatt** [comment]
Yep, that's a bug.  We already discard output in commands like Describe, but looks like that's not being done out where test scripts themselves get invoked.

**Comment by mikeclayton** [comment]
Hi Dave,

Thanks for confirming.

I've added some tests around this behaviour (which initially failed as expected), and what looks like a deceptively simple change to fix the new tests. All the Pester pass locally (except half a dozen do to with tagging and versioning) but I might have missed something.

See https://github.com/pester/Pester/compare/master...mikeclayton:feature/1060-capture-test-output for commits - I can send a PR if it looks ok (or if you want one here to comment on).

Note - I found 2 ways of piping an array into Should in the new test - let me know if you have a preference which one to use.

Cheers,

M

**Comment by it-praktyk** [comment]
@mikeclayton,
it took me some time to understand where the issue is. Finally, I've tested the code updated by you using a file with the content.

```powershell
"Loading DemoFunction1 - Write-Output" | Write-Output

"Loading Demofunction1 - Out-Host" | Out-host

function DemoFunction1 {

    param (

        [Int]$FirstParam

    )

    "Provided parameter $FirstParam" | Write-Output

    "Provided parameter $FirstParam" | Write-Output

    Return $FirstParam

}
```
and [tests](https://github.com/equelin/Format-Pester/blob/master/demo/DemoFunction1.Tests.ps1).

I confirm that it works as expected now. It means that output from other commands is not included in the object returned when the `PassThru` parameter is used. 

I've tested the code updated by you using PowerShell 2.0 and it works fine too.

Feel free to make the new pull request.

Thank you for your contribution.
